pkg/apis/pipeline/v1/matrix_types.go

```diff
@@ -300,29 +300,17 @@ func (m *Matrix) validateCombinationsCount(ctx context.Context) (errs *apis.Fiel
 	return errs
 }
 
-// validateParams validates the type of Parameter for Matrix.Params and Matrix.Include.Params
-// Matrix.Params must be of type array. Matrix.Include.Params must be of type string.
-// validateParams also validates Matrix.Params for a unique list of params
-// and a unique list of params in each Matrix.Include.Params specification
-func (m *Matrix) validateParams() (errs *apis.FieldError) {
+// validateUniqueParams validates there are no duplicate parameters
+// within Matrix.Params or Matrix.Include Params
+func (m *Matrix) validateUniqueParams() (errs *apis.FieldError) {
 	if m != nil {
 		if m.HasInclude() {
 			for i, include := range m.Include {
 				errs = errs.Also(include.Params.validateDuplicateParameters().ViaField(fmt.Sprintf("matrix.include[%d].params", i)))
-				for _, param := range include.Params {
-					if param.Value.Type != ParamTypeString {
-						errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("parameters of type string only are allowed, but got param type %s", string(param.Value.Type)), "").ViaFieldKey("matrix.include.params", param.Name))
-					}
-				}
 			}
 		}
 		if m.HasParams() {
 			errs = errs.Also(m.Params.validateDuplicateParameters().ViaField("matrix.params"))
-			for _, param := range m.Params {
```

Comment:
isn't `$(params.bar[*])` still an array param?

Reply:
It is of type string parameter referencing an array param

Reply:
Interesting, I'm curious what error is returned here, and if it would make sense to address that instead? (maybe the answer is no, but this is surprising behavior at least to me)

Reply:
Moved to the reconciler

Reply:
Can you elaborate on whether it would make sense to treat this as an array param? (I suspect the answer is no but would like to have a better understanding of the problem this is addressing.)

Reply:
This is the same syntax used to define whole array result references 
https://github.com/tektoncd/pipeline/blob/a0b5cc55986818bff3c068f85e440e1236f172d1/examples/v1beta1/pipelineruns/beta/pipelinerun-array-results-substitution.yaml#L32

Reply:
Sorry, I'm still a bit confused, let me rephrase.

It sounds like you're saying the webhook treats the string "$(params.foo[*])" in the following example as a string param, but the reconciler parses it and treats it as an array param. Is that correct?

```
params:
- name: foo
   value: $(params.foo[*])
```

Your PR description says "Prior to this, validation fails because a string referencing an array type was considered a string type." Is there any issue or comment where I can see an example of this, and the error message that is returned?

My question is whether it makes sense to parse this string in the webhook, and have both the webhook and reconciler treat it as an array param, as opposed to the approach taken in this PR (moving validation to the reconciler). This is a genuine question, I really don't know what the answer is here so I'm curious to hear your opinion.

Reply:
Yes, this is mentioned here: https://github.com/tektoncd/pipeline/pull/6056#issuecomment-1406883457

I think with the current logic, it makes sense to move the validation to the reconciler since the replacements are being made in the reconciler, but am open to other ideas. Do you have any cons to this approach?

cc: @jerop

Reply:
ah ok sorry it took me so long to understand this 😓 in general validating in the webhook fails faster than validating in the reconciler, but I think you're right that moving this to the reconciler is the right approach here. Thanks for bearing with me!